db/models/aggregate_functions.py

```diff
@@ -153,9 +154,35 @@ def submit_application(application_id):
 def update_form(application_id, form_name, question_json):
     try:
         form_sql_row = FormsMethods.get_form(application_id, form_name)
-        form_sql_row.json = question_json
-        db.session.commit()
-        update_statuses(application_id, form_name)
-        return form_sql_row.as_json()
+        # Running update form for the first time
+        if question_json and not form_sql_row.json:
+            application_edited(application_id, question_json, form_name)
+        # Removing all data in the form (should not be allowed)
+        elif form_sql_row.json and not question_json:
+            current_app.logger.error(
+                "Application update aborted for application_id:"
+                f" '{application_id}. Invalid data supplied"
+            )
+            raise Exception("ABORTING UPDATE, INVALID DATA GIVEN")
+        # Updating form subsequent times
+        elif (
+            form_sql_row.json
+            and form_sql_row.json[0]["fields"] != question_json[0]["fields"]
+        ):
+            application_edited(application_id, question_json, form_name)
     except sqlalchemy.orm.exc.NoResultFound as e:
         raise e
+    db.session.commit()
+    return form_sql_row.as_json()
+
+
+def application_edited(application_id, question_json, form_name):
```

Comment:
nice refactoring! should this function be renamed to account for your new additions, maybe something like update 'update_form_data'

Reply:
can change it to  update_application_data? i think that's more accurate as we're only changing data stored in the application object/ record?

Reply:
I think were changing form row also on line 183

Reply:
ahhh silly me 😆  maybe something like application_form_updated?